Allow passing `variables` via raw YAML

nkaretnikov opened the pull request:
Fixes #350.

<!-- Reference the issue corresponding to this PR. If an issues does not exist, consider opening one or writing a detailed description for what this PR changes and it's value in the following sections. -->
<!-- GitHub Docs on Keywords: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests -->

## Description
<!-- What is the purpose of this pull request? -->

This pull request:
- is a companion PR to https://github.com/conda-incubator/conda-store/pull/721
- allows passing `variables` via raw YAML, see:
https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-virtual.html#overriding-detected-packages
- doesn't create new GUI elements (like the ones we have for channels and dependencies) because the conda-store server currently supports only `CONDA_OVERRIDE_CUDA` due to a conda-lock limitation (GUI elements can be added later once we have a chance to discuss how that should look like).

Tested manually:
- [x] `variables` is present and equal to `{}` when creating an environment
- [x] `variables` can be edited when creating an environment and the result can be later seen again via the new or admin UI
- [x] `variables` can be edited after the environment was created, the new result is saved and can be seen again via the new or admin UI (as part of the new build)
- [x] the server receives the provided `variables` value in `action_solve_lockfile` (in all of the above cases).

## Pull request checklist
<!-- Quick checklist to ensure high-quality Pull Request. -->

- [x] Did you test this change locally?
- [ ] Did you update the documentation (if required)?
- [x] Did you add/update relevant tests for this change (if required)?

## Additional information
<!-- Do you have any other information about this pull request? This may include screenshots, references, and/or implementation notes. -->

These tests should pass locally:

```
yarn run test test/environmentDetails/SpecificationEdit.test.tsx --coverage=false
yarn run test test/environmentCreate/SpecificationCreate.test.tsx --coverage=false
```

nkaretnikov commented on src/features/environmentCreate/components/Specification/SpecificationCreate.tsx:
```diff
@@ -59,21 +66,27 @@ export const SpecificationCreate = ({ onCreateEnvironment }: any) => {
+    dependencies: string[],
+    variables: Record<string, string>
+  ) => {
+    return stringify({ channels, dependencies, variables });
```
I've removed custom pretty-printing code that we had before. I don't see much value in it because:
- it will be ignored if either value is not empty since it uses `&&`
- with the newly-added `variables` field, the number of possible empty/non-empty variations is even bigger
- what `stringify` returns is valid YAML and I think it's readable.

You can see the old version in the diff.

The new version looks like this:
```yaml
channels: []
dependencies: []
variables: {}
```

nkaretnikov replied:
I probably should replace this with `stringify` wherever this is called because we're doing this in `src/features/environmentDetails/components/Specification/SpecificationEdit.tsx` anyway.

nkaretnikov commented on src/features/variables/variablesSlice.ts:
```diff
@@ -0,0 +1,40 @@
+import { createSlice } from "@reduxjs/toolkit";
```
This file here is essentially the same as `src/features/channels/channelsSlice.ts`. I've just replaced the variable name and type.

nkaretnikov commented on src/store/selectors/condaSpecificationSelector.ts:
```diff
@@ -7,14 +7,16 @@ export const selectCondaSpecification = (
   const {
     channels: { channels },
     requestedPackages: { requestedPackages },
+    environmentCreate: { variables },
```
FIXME: this is a mistake. It should use the global `variables` state.

nkaretnikov commented on src/store/selectors/condaSpecificationSelector.ts:
```diff
@@ -7,14 +7,16 @@ export const selectCondaSpecification = (
     name,
-    prefix
+    prefix,
+    variables: variables
```
Nit: just use `variables` since the name is the same in both cases.

nkaretnikov commented on src/features/environmentCreate/components/Specification/SpecificationCreate.tsx:
```diff
@@ -99,7 +104,7 @@ export const SpecificationCreate = ({ onCreateEnvironment }: any) => {
         {show ? (
           <CodeEditor
-            code={formatCode(channels, requestedPackages)}
+            code={stringify(channels, requestedPackages, variables)}
```
See https://github.com/conda-incubator/conda-store-ui/pull/354#discussion_r1444104490 for an explanation.

nkaretnikov replied:
Okay, this was a wrong decision. When creating a new env with a clean DB, this would be just `[]` which is not very user-friendly. I'll revert.

nkaretnikov commented on src/features/environmentCreate/components/Specification/SpecificationCreate.tsx:
```diff
@@ -59,21 +66,34 @@ export const SpecificationCreate = ({ onCreateEnvironment }: any) => {
+      return "channels:\n  -\ndependencies:\n  -\nvariables: {}";
     }
-    return stringify({ channels, dependencies: dependencies });
+    return stringify({ channels, dependencies, variables });
```
How the custom pretty-printer looks when no data is there:

<img width="208" alt="Screen Shot 2024-01-08 at 03 43 00" src="https://github.com/conda-incubator/conda-store-ui/assets/10469885/881be8fb-dcc2-4235-a228-8d413efcec17">

gabalafou reviewed (commented):
What is the urgency on this? Would it be too much to ask for some tests?

I'm just thinking that the repo already has a nice base of test files, and you already wrote up a set of manual tests, so maybe it wouldn't be too hard to turn some of those into automated tests. What do you think?

gabalafou commented on src/features/environmentDetails/components/Specification/SpecificationEdit.tsx:
```diff
@@ -116,7 +131,7 @@ export const SpecificationEdit = ({
       dispatch(updateChannels(code.channels));
     } else {
-      setCode({ dependencies: requestedPackages, channels });
+      setCode({ dependencies: requestedPackages, variables, channels });
```
Why is there no dispatch line for variables in the if-branch just above?

nkaretnikov replied:
Because you can only set variables by editing the YAML specification. There's no dedicated UI element for variables, which is what the UI block above is for. See the attached screenshots:

<img width="894" alt="Screen Shot 2024-01-31 at 18 17 49" src="https://github.com/conda-incubator/conda-store-ui/assets/10469885/ea8e78c2-42d0-4d06-942b-c19bf19c9c24">
vs.
<img width="901" alt="Screen Shot 2024-01-31 at 18 17 40" src="https://github.com/conda-incubator/conda-store-ui/assets/10469885/051f11e6-2ce1-4630-be6e-d1b263e3a9a5">

gabalafou replied:
Yes, I thought that might be the case. However, the thing that's confusing is that `dispatch` is about updating the global store, which isn't UI.

So even though there isn't a dedicated GUI element to surface the state changes made to the `variables` key in the YAML editor, you should still update the app state (unless I'm still missing something here). Why? Imagine that somebody else comes into the codebase whose task is to add some component that displays the value of `variables`. They understand that `variables` can be updated somehow, but how it's updated is not really their concern. They're just tasked with writing the component that displays the value of `variables`. So they would hook into the global store, read the `variables` key, and display the values. But if the code for the YAML editor never updates the global store (via `dispatch`), then our programmer has just coded up a completely useless component that never displays the value of `variables`.

As an aside, it looks to me that the calls to `dispatch` are in an odd place. I wonder why `dispatch` isn't called on editor update.

nkaretnikov replied:
The actionable part of my reply: I think the GUI parts can be implemented separately. This was intended as a quick fix for a problem that some users were having. More details below, feel free to skip.

----

#### How state updates work

The way this currently works:
- if you make a change to the raw YAML, the variables are added to the app state
- if you make a change via the GUI, the current variables state is used (could be the initial empty state, could be the one set via raw YAML before).

#### Why updating state in raw YAML seems fine

To me, this seems better than updating variables' state on GUI updates (which would need to remove previously set values since we have no GUI elements for variables at the moment). Using GUI is convenient for adding packages. Setting variables is not something that should happen often. I also wanted to make as few changes as possible to keep this well-scoped, so this fits nicely into existing logic.

#### Why variables are needed

The variables support was supposed to be a quick fix to unblock Nebari users. The initial version of this PR was ready right after it was implemented on the server. It's just that we had no one to review the UI part for a bit.

On Nebari, users usually spin up CPU-only instances, but they often want to build environments with GPU support anyway, in case that's needed later. Also, conda-store-server currently supports setting only one particular variable (`CONDA_OVERRIDE_CUDA`) and ignores everything else.

#### Process for GUI changes

When GUI changes are required, our process has been to get a design from our designer first, to ensure the best UX. Then someone implements that on the UI side. That's why I only implemented the non-GUI parts of this.

#### Process for avoiding errors

Regarding potential issues if GUI were to be implemented separately: we have a process in place where each PR requires testing and is reviewed by someone on the team with relevant experience.

gabalafou reviewed (commented):
I haven't tested any of this, but I carefully reviewed the code. It looks great. I only spotted one thing, and I left an inline question.

But I also have a more general question. Are we sure about the word choice for `variables`? I'm concerned that the name might be a little too generic, and we may come to regret the word choice later.

trallard commented:
> But I also have a more general question. Are we sure about the word choice for variables? I'm concerned that the name might be a little too generic, and we may come to regret the word choice later.

Agree with you here @gabalafou we should be favouring descriptive variable names, so I suggest something like `conda_env_variable` or the such @nkaretnikov

trallard commented on src/features/environmentCreate/components/Specification/SpecificationCreate.tsx:
```diff
@@ -16,14 +16,15 @@ import { getStylesForStyleType } from "../../../../utils/helpers";
 export const SpecificationCreate = ({ onCreateEnvironment }: any) => {
   const dispatch = useAppDispatch();
-  const { channels, requestedPackages } = useAppSelector(
+  const { channels, requestedPackages, variables } = useAppSelector(
```
See comments in PR regarding the naming choice

nkaretnikov commented:
This is now blocked because CI is failing (unless this is urgently needed and we want to revert to my original version and merge without CI). The reason is explained in https://github.com/conda-incubator/conda-store-ui/pull/360#issuecomment-1923360060.

nkaretnikov commented:
@trallard @gabalafou I renamed variables where possible and added new tests. PTAL.

About variable names: the choice of names matter or it can result in subtle bugs due to serialization. We want to make sure that we always generate `variables` when we serialize.

I ended up following the conventions of existing code: essentially, we need to use `variables` where `dependencies` is used and can use a more verbose name in places where `requestedPackages` is used. This is a good rule of thumb because it's not often clear that a certain object may be serialized later.

This is due to how JavaScript objects are serialized. Note how the key is created automatically here based on the variable name:

```js
> var foo = 'bar'
undefined
> JSON.stringify({foo})
'{"foo":"bar"}'
```

gabalafou commented on src/features/environmentDetails/components/Specification/SpecificationEdit.tsx:
```diff
@@ -116,7 +142,11 @@ export const SpecificationEdit = ({
       dispatch(updateChannels(code.channels));
     } else {
-      setCode({ dependencies: requestedPackages, channels });
+      setCode({
```
So I tested this manually and now I believe pretty strongly that we should fix the asymmetry I mentioned before and dispatch the `updateEnvironmentVariables` action when the user toggles from YAML to GUI mode, just as we do for channels and dependencies; otherwise, the user loses the changes they made in YAML mode.

Steps to recreate. Go to an environment, click edit. Toggle into YAML mode. Change the `variables` field. Toggle into GUI mode. Toggle back into YAML mode. Notice that the changes you made to the `variables` field are undone.

nkaretnikov replied:
> Steps to recreate. Go to an environment, click edit. Toggle into YAML mode. Change the variables field. Toggle into GUI mode. Toggle back into YAML mode. Notice that the changes you made to the variables field are undone.

Oh, that's a good point! I had no idea it works like that. I thought the whole reason we have the Save button there is to save the changes. But yeah, I've just tried this with dependencies and it does update them in the GUI simply if you switch back and forth between YAML and GUI without clicking on Save.

gabalafou reviewed (commented):
If we really need to get this PR deployed before doing any fix-ups, then I am willing to approve it, but otherwise, I really think there are two things that need to happen before it gets merged:

1. Edits to variables should be dispatched (see inline comment)
2. The test code needs to be rewritten (there are many things wrong with the added tests, it will save me a lot of time if we could review them together synchronously)

nkaretnikov commented:
> The test code needs to be rewritten (there are many things wrong with the added tests, it will save me a lot of time if we could review them together synchronously)

Scheduled a meeting to discuss this.

trallard commented:
Thanks for your review @gabalafou, as usual it is a critical and in depth one 
I agree with having the code and tests fixed before merging.

gabalafou commented on test/environmentCreate/SpecificationCreate.test.tsx:
```diff
@@ -48,20 +71,40 @@ describe("<SpecificationCreate />", () => {
     const switchButton = component.getByLabelText("YAML", { exact: false });
     fireEvent.click(switchButton);
 
+    expect(
+      screen.getByText((content, element) => {
+        return (
+          element?.textContent === "channels:  -dependencies:  -variables: {}"
+        );
+      })
+    );
+
```
```suggestion
    const textbox = await component.findByText("channels: -dependencies: -variables: {}");
    expect(textbox).toBeInDocument();
```

I assume this will give a better error message.

JavaScript developers are not used to seeing an unchained/naked expect() call (e.g. `toBeInDocument`). 

A JS dev would have to know that `getByText()` throws an error, and that's essentially what will cause this test to fail, rather than the chained call to expect() failing (and throwing).

For reference, [testing-library about queries](https://testing-library.com/docs/queries/about#types-of-queries)

nkaretnikov replied:
I'll try to rewrite these using the suggestion above, but here's some context for why the current code is written the way it is.

First, this screenshot, which I took while developing this, shows how the HTML code looks like (this is the case when `variables` is empty, set to `{}`):

<img width="332" alt="Screen Shot 2024-02-11 at 16 25 57" src="https://github.com/conda-incubator/conda-store-ui/assets/10469885/4080aed5-fe5b-4415-bc32-4778f66a05c5">

So my initial idea was to search for a substring (like `variables`, which is inside the `span` tag) and then look at the parent node, which is inside the `div`, and check the string there. However, I then discovered this discussion:

https://stackoverflow.com/questions/62770973/parent-node-in-react-testing-library

with this top answer:

```js
screen.getByText("span text").closest("div")
```

What I don't like about this is that this doesn't establish a relationship to the element I'm matching on. It's just the closest div, which could be elsewhere.

There's also a discussion there on what's considered good practice in the react testing library and people say there are even default eslint rules to enforce certain things. And it might be not a good idea to access parent elements directly.

So I wanted to find something else that would allow me to capture the entire textual output of the YAML.

The only thing I found was `textContent`. But how can I get that? I looked for a suggestion again and came across the `getByText((content, element) => ...` overload, probably in this discussion:

https://stackoverflow.com/questions/61675320/how-to-get-the-child-div-of-parent-div-with-some-test-id-using-react-testing-lib

So I just ended up using this with `textContent`.

One thing I definitely need to improve is to match on `component` instead of the entire document.

----

@gabalafou While I was typing this reply, I actually found another suggestion here:

https://stackoverflow.com/questions/55509875/how-to-query-by-text-string-which-contains-html-tags-using-react-testing-library

So I wonder if this would be better (it still allows to match on `textContent` and takes advantage of the `.to` methods of the testing library):

```js
const el = getByText('Name:', { exact: false })
expect(el.textContent).toEqual('Name: Bob (special guest)');
```

gabalafou commented on test/environmentCreate/SpecificationCreate.test.tsx:
```diff
@@ -33,6 +35,27 @@ describe("<SpecificationCreate />", () => {
+          "channels: []dependencies:  - python>5.0  - numpyvariables: {}"
+        );
+      })
+    );
```
Please move this into its own test.

nkaretnikov replied:
Context for this: we discussed how the tests should match their description. The problem here is that I inserted code that does more than just checking that switching to YAML works. So I need to move this new logic into a standalone test (but copy-pasting the current code is fine).

nkaretnikov commented:
We had a call about tests in this PR. The comments above are from that call, but there's one issue we didn't have time to get to:

@gabalafou raised an issue about us comparing `textContent` to raw strings in the testsuite, because it might potentially break later when the rendering logic changes (such as when we update a dependency or refactor the code), since the strings include whitespace. Gabriel wanted to see the code that generates these on the UI side to better understand how this should be checked instead.

My answer to this is as follows: _checking the raw output is the whole point of the test!_ From this standpoint, it's irrelevant how the current code is written. What I'm trying to avoid is potential future problems when someone updates the logic in the codebase or some dependency breaks (someone might add a new field or remove a field or change a list to a dict or something else). This format is what the users will see and it needs to match the conda format as well. Ideally, I'd like this tests to include newlines as well, but it's not possible when using `textContent` because newlines are rendered via CSS (the entries in the YAML text field are `div` and `span` elements with attributes).

As for these tests breaking because some dependency started to format lists with 4 spaces by default instead of 2, I'm not really worried about that. It'd be easy to update. However, I think there's real value in detecting formatting/generation errors early as opposed to checking other parts of the system (the formatting issue here might manifest as an input issue in conda or it might just be ignored, but the users will see strange output in the UI, which doesn't really match the conda spec). And yes, ideally, I'd include newlines as well, but it's tricky to do because of how this is rendered (see the comment about CSS above). So the current approach is the second best thing to getting 1-to-1 raw YAML output out of the system. I'll also add that this is how the current playwright tests are written as well, see `test/playwright/test_ux.py`:

```py
page.get_by_text("channels: - conda-forgedependencies: - rich - pip: - nothing - ipykernel").fill("channels:\n  - conda-forge\ndependencies:\n  - rich\n  - python\n  - pip:\n      - nothing\n  - ipykernel\n\n")
```

That said, I'd be happy to reconsider and welcome suggestions! I'm just not sure how else we can write these to catch potential future changes to the generated YAML.

nkaretnikov commented:
I had another meeting about this with @gabalafou. We discussed tests comparing against raw strings.

I think I'll just try to rewrite these tests such that we check: `if <key>, then <value> == <expected>` and try to do it in a way that ignores newlines/whitespace.

This will address the point that Gabriel made about this potentially breaking for unrelated reasons, while satisfying what I want to test for, which is that variables have very specific names.

I still think we should find a way to capture the entire output, newlines included, and test that as well, somewhere. But it's something we can postpone for later.

nkaretnikov commented:
@gabalafou

_Note: I still need to fix the issue where variables are not saved when you switch from/to YAML. I'll do so shortly._

But I've updated the tests based on our discussions. PTAL.

The reason I'm using `textContent ... toBe` is because it matches the string end-to-end. If you prefer strings not to have leading whitespace at all, another option is to use `.toHaveTextContent`.

Compared with before, these tests now print the HTML node value when it fails to match. They also look at a single component instead of searching through the entire page.

I'm also attaching a screenshot, which shows how the HTML looks like when the YAML file has dependencies. It explains why I'm using `nextElementSibling` and `closest`. If you can think of a better way to access this information, I welcome your feedback.

<img width="549" alt="Screen Shot 2024-02-22 at 20 19 17" src="https://github.com/conda-incubator/conda-store-ui/assets/10469885/b099bdd4-bb86-4b32-940c-0ca05d85f643">

nkaretnikov commented:
The commits I pushed yesterday add support for (temporary) saving the state of variables when switching between GUI/YAML, which is another issue that was highlighted as part of this review. I think this addresses all the feedback I've received so far.

gabalafou reviewed (changes requested):
It seems that this PR is blocking the Conda Store release. Here's what needs to happen for me to feel good about approving this PR:

1. Remove the on render update (see inline comment)
2. Remove changes to the test files. From my experience coding and testing front-end apps, these tests are far too prone to breaking. When I asked for tests, I was asking for something much simpler.

nkaretnikov commented:
Gabriel and I had a very productive discussion about this. To get this accepted:
- Remove all changes to tests because these will be flaky and we don't have good testing infrastructure for those types of tests
- Remove extraneous dispatch calls (only a couple of those actually matter for saving the state).

nkaretnikov commented:
@gabalafou PTAL. I'll rebase before landing. Do not want to rebase right now to avoid making the diff hard to read.

gabalafou reviewed (approved):
Thanks @nkaretnikov!

nkaretnikov commented:
@trallard this is ready, but it needs your approval before I can merge.

gabalafou commented on src/features/environmentCreate/components/Specification/SpecificationCreate.tsx:
```diff
@@ -59,27 +84,41 @@ export const SpecificationCreate = ({ onCreateEnvironment }: any) => {
+  ) => {
     if (channels.length === 0 && dependencies.length === 0) {
-      return "channels:\n  -\ndependencies:\n  -";
+      // Note: these empty pretty-printed lists translate to [null]
```
Nevermind, I see the comment on line 49 above.
